modules/lang-painless/src/main/java/org/elasticsearch/painless/lookup/PainlessLookupUtility.java

```diff
@@ -19,264 +19,284 @@
+                throw new IllegalArgumentException("type [" + canonicalTypeName + "] not found", cnfe);
+            }
         }
 
-        return javaType;
+        throw new IllegalArgumentException("type [" + canonicalTypeName + "] not found");
     }
 
-    public static Class<?> painlessDefTypeToJavaObjectType(Class<?> painlessType) {
-        if (painlessType.isArray()) {
-            Class<?> painlessTypeComponent = painlessType.getComponentType();
-            int arrayDimensions = 1;
+    public static String typeToCanonicalTypeName(Class<?> type) {
+        Objects.requireNonNull(type);
 
-            while (painlessTypeComponent.isArray()) {
-                painlessTypeComponent = painlessTypeComponent.getComponentType();
-                ++arrayDimensions;
-            }
+        String canonicalTypeName = type.getCanonicalName();
 
-            if (painlessTypeComponent == def.class) {
-                char[] asmDescriptorBraces = new char[arrayDimensions];
-                Arrays.fill(asmDescriptorBraces, '[');
+        if (canonicalTypeName.startsWith(def.class.getName())) {
+            canonicalTypeName = canonicalTypeName.replace(def.class.getName(), DEF_TYPE_NAME);
+        }
 
-                String asmDescriptor = new String(asmDescriptorBraces) + Type.getType(Object.class).getDescriptor();
-                Type asmType = Type.getType(asmDescriptor);
+        return canonicalTypeName;
+    }
 
-                try {
-                    return Class.forName(asmType.getInternalName().replace('/', '.'));
-                } catch (ClassNotFoundException exception) {
-                    throw new IllegalStateException("internal error", exception);
-                }
+    public static String typesToCanonicalTypeNames(List<Class<?>> types) {
+        StringBuilder typesStringBuilder = new StringBuilder("[");
+
+        int anyTypesSize = types.size();
+        int anyTypesIndex = 0;
+
+        for (Class<?> painlessType : types) {
+            String canonicalTypeName = typeToCanonicalTypeName(painlessType);
+
+            typesStringBuilder.append(canonicalTypeName);
+
+            if (++anyTypesIndex < anyTypesSize) {
+                typesStringBuilder.append(",");
             }
-        } else if (painlessType == def.class) {
-            return Object.class;
         }
 
-        return painlessType;
-    }
+        typesStringBuilder.append("]");
 
-    public static String anyTypeNameToPainlessTypeName(String anyTypeName) {
-        return anyTypeName.replace(def.class.getName(), DEF_PAINLESS_CLASS_NAME).replace('$', '.');
+        return typesStringBuilder.toString();
     }
 
-    public static String anyTypeToPainlessTypeName(Class<?> anyType) {
-        if (anyType.isLocalClass() || anyType.isAnonymousClass()) {
-            return null;
-        } else if (anyType.isArray()) {
-            Class<?> anyTypeComponent = anyType.getComponentType();
+    public static Class<?> javaTypeToType(Class<?> javaType) {
```

Comment:
I'm unsure what this means. Can you elaborate, maybe with a javadoc string?

Reply:
Went through all the methods in PainlessLookupUtility and added some brief javadocs for each.  Will need more passes, but it at least should clear up some of the more confusing aspects of some of these methods.